occa: add v2.0.0, rewrite package for CMake support

thilinarmtb opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
Update OCCA package to use a CMake build and add a new release.
Also, we removed support for some of the older versions of OCCA.

thilinarmtb committed:
occa: Add v2.0.0 to the spack package

thilinarmtb committed:
occa: Expose HIP backend in spack builds

thilinarmtb committed:
occa: Add a test to see if the occa binary works

thilinarmtb committed:
occa: Switch to a CMake build

thilinarmtb commented:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
@spackbot fix style

thilinarmtb commented:
@v-dobrev : Just realized that I may be breaking CEED 1.0, 2.0 and 3.0 packages.

thilinarmtb commented:
Is there a way in spack to use both `make` and `cmake` in a given package? OCCA versions 
`<1.1.0` uses `make` while the newer versions use `CMake`. Can we have select the build
system based on the version?

tgamblin commented:
> Is there a way in spack to use both make and cmake in a given package? 

Yes. See https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems

Look at, e.g., `cp2k` or try:
```
> spack pkg grep -l MakefileBuilder | xargs grep CMakeBuilder
```

to see others that do this.

thilinarmtb commented:
> > Is there a way in spack to use both make and cmake in a given package?
> 
> Yes. See https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems
> 
> Look at, e.g., `cp2k` or try:
> 
> ```
> > spack pkg grep -l MakefileBuilder | xargs grep CMakeBuilder
> ```
> 
> to see others that do this.

This is exactly what I was looking for ! Thank you very much !

thilinarmtb committed:
occa: Select build system based on OCCA version

thilinarmtb commented:
@spackbot fix style

thilinarmtb committed:
occa: Set build defaults for all the variants

And turn-off all the variants by default.

thilinarmtb committed:
occa: Move test execution to the builder class

thilinarmtb committed:
occa: Fix a typo and update comments

v-dobrev reviewed (dismissed):
Look good to me. However, I did not test the changes. @thilinarmtb, did you test older versions?

thilinarmtb commented:
> Look good to me. However, I did not test the changes. @thilinarmtb, did you test older versions?

I tested the version from 2.0.0 to 1.0.9. Will do a bit more testing before merging this.

thilinarmtb committed:
occa: Setup HIP variables correctly

thilinarmtb committed:
occa: Turn-off OpenCL due to build errors

thilinarmtb committed:
occa: Enable OpenMP when version >= 1.1.0

thilinarmtb committed:
occa: Adjust build system switch logic

thilinarmtb committed:
occa: Add conflicts for +openmp and +rocm

Also, patch v1.2.0 for a compilation error.

thilinarmtb commented:
> Look good to me. However, I did not test the changes. @thilinarmtb, did you test older versions?

I tested all the versions on Frontier at OLCF. Will do the same when Polaris becomes available at ANL.

thilinarmtb commented:
Tested all the versions successfully on Polaris as well.

thilinarmtb committed:
occa: Add missing dependency on gmake

v-dobrev reviewed (dismissed):
I assume the latest force-push is equivalent to the earlier version that you tested @thilinarmtb?

thilinarmtb commented:
> I assume the latest force-push is equivalent to the earlier version that you tested @thilinarmtb?

Yes, I rebased over the `develop` and added `gmake` as a dependency. I will do a quick test run 
again on Frontier, Aurora and Polaris.

alalazo commented on var/spack/repos/builtin/packages/occa/package.py:
```diff
@@ -33,91 +38,177 @@ class Occa(Package):
+    build_system(
+        conditional("cmake", when="@1.1.0:"),
+        conditional("makefile", when="@:1.0.9"),
+        default="makefile",
```
This default is inconsistent with versions defaults. Can we default to `cmake`, or is there a reason for this?

thilinarmtb replied:
Older versions of OCCA (<=1.0.9) doesn't support `CMake` where as `make` is supported on all the versions.

alalazo replied:
This is not what you say in the directive:
```
conditional("makefile", when="@:1.0.9"),
```
This creates a condition for clingo where the optimal version has a suboptimal variant and vice-versa. It's much better if all defaults are consistent.

thilinarmtb replied:
I was hoping what I have above will make sure that it will use `CMake` for versions `>=1.1.0` 
and `make` for versions `<=1.0.9`. A default version may not be necessary in this case
since above two conditions covers all the cases.

But seems like what you suggest is the following?

```python
build_system(conditional("cmake", when="@1.1.0:"), "makefile", default="cmake")
```

alalazo replied:
What I am suggesting is:
```suggestion
        default="cmake",
```
I guess, whether you want to remove the condition on the `makefile` build-system is up to you, as a package maintainer.

alalazo commented on var/spack/repos/builtin/packages/occa/package.py:
```diff
@@ -33,91 +38,177 @@ class Occa(Package):
-        make("-C", prefix)
+    patch("occa-v1.2.0-cstdint.patch", when="@1.2.0")
+
+    @run_after("install")
+    def check_install(self):
+        version_cmd = join_path(self.prefix.bin, "occa") + " version"
+        val = os.system(version_cmd)
+        if val != 0:
+            raise RuntimeError("Calling `occa version` failed !")
```
Can we use the `Executable` class here? Also, shouldn't this be encoded as a test? Wouldn't something ilke `make test` cover this simple case?

thilinarmtb replied:
We don't run OCCA tests as they depend on the GPU computing runtime to be 
available (which sometimes is not available on the build machine). Are you referring 
to the `Executable` class [here](https://spack.readthedocs.io/en/latest/spack.util.html#spack.util.executable.Executable)?

alalazo replied:
Yes. You can grep other packages. That class is imported in the recipe via `from spack.package import *`

alalazo commented on var/spack/repos/builtin/packages/occa/package.py:
```diff
@@ -33,91 +38,177 @@ class Occa(Package):
+            cuda_libs = find_libraries(cuda_libs_list, cuda_dir, shared=True, recursive=True)
+            include_dirs.append(cuda_dir.include)
+            library_dirs.extend(cuda_libs.directories)
+            args.append("-DOCCA_ENABLE_CUDA=ON")
+        else:
+            args.append("-DOCCA_ENABLE_CUDA=OFF")
```
Here we can use `self.define_from_variant` We can also use `self.define` below instead of raw strings.

thilinarmtb replied:
Thanks ! Fixed in the latest commit.

alalazo commented on var/spack/repos/builtin/packages/occa/package.py:
```diff
@@ -33,91 +38,177 @@ class Occa(Package):
+
+        return args
+
+    @run_after("build")
+    @on_package_attributes(run_tests=True)
+    def check(self):
+        if self.pkg.run_tests:
+            make("-C", self.prefix, "test", parallel=False)
```
:thinking: Is this needed because of `parallel=False`?

thilinarmtb replied:
It is not clear to me what `this` refers to. Is it the entire `make` command or a part of it?

alalazo replied:
Having to override the function. The default implementation is, for CMake:
```python
    def check(self) -> None:
        """Search the CMake-generated files for the targets ``test`` and ``check``,
        and runs them if found.
        """
        with fs.working_dir(self.build_directory):
            if self.generator == "Unix Makefiles":
                self.pkg._if_make_target_execute("test", jobs_env="CTEST_PARALLEL_LEVEL")
                self.pkg._if_make_target_execute("check")
            elif self.generator == "Ninja":
                self.pkg._if_ninja_target_execute("test", jobs_env="CTEST_PARALLEL_LEVEL")
                self.pkg._if_ninja_target_execute("check")
```
so, to make it run serially, it seems you just need to set `CTEST_PARALLEL_LEVEL`?

alalazo commented on var/spack/repos/builtin/packages/occa/package.py:
```diff
@@ -33,91 +38,177 @@ class Occa(Package):
+        install_tree(".", prefix)
+        make("-C", prefix)
+
+    @run_after("build")
+    @on_package_attributes(run_tests=True)
+    def check(self):
+        make("-C", self.prefix, "test", parallel=False)
```
Same question as above

thilinarmtb committed:
occa: Use self.define and self.define_from_variant

thilinarmtb committed:
occa: Wrap `occa` binary using `Executable`

thilinarmtb committed:
occa: Set `CTEST_PARALLEL_LEVEL=1`

thilinarmtb committed:
occa: Default to `cmake` build system

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

thilinarmtb committed:
occa: Set the environment variable value correctly

thilinarmtb commented:
I tested the latest changes on Aurora, Polaris and Frontier. I think we can go ahead and merge the PR once the CI passes.
